External DRS URIs in lm10 snapshots break `/repository/files`

Traceback
```
[ERROR]	2025-12-19T18:38:47.136Z	3471cc03-5124-4299-b7d5-c51eec4dbc32	azul-service-dev	Caught exception for path /fetch/repository/files/{file_uuid}
Traceback (most recent call last):
  File ""/var/task/chalice/app.py"", line 1922, in _get_view_function_response
    response = view_function(**function_args)
  File ""/var/task/app.py"", line 1534, in fetch_repository_files
    body = _repository_files(file_uuid, fetch=True)
  File ""/var/task/app.py"", line 1548, in _repository_files
    return app.repository_controller.download_file(catalog=catalog,
           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^
                                                   fetch=fetch,
                                                   ^^^^^^^^^^^^
    ...<2 lines>...
                                                   headers=headers,
                                                   ^^^^^^^^^^^^^^^^
                                                   authentication=request.authentication)
                                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File ""/var/task/azul/service/repository_controller.py"", line 288, in download_file
    download.update(plugin, authentication)
    ~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^
  File ""/var/task/azul/plugins/repository/tdr.py"", line 274, in update
    access = drs_client.get_object(self.file.drs_uri,
                                   access_method=AccessMethod.gs)
  File ""/var/task/azul/drs.py"", line 362, in get_object
    return self._get_object(drs_uri, access_method)
           ~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^
  File ""/var/task/azul/drs.py"", line 384, in _get_object
    return self._get_object_access(drs_uri, access_id, AccessMethod.https)
           ~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File ""/var/task/azul/drs.py"", line 427, in _get_object_access
    raise DRSStatusException(url, response)
azul.drs.DRSStatusException: ('Unexpected response from https://gen3.biodatacatalyst.nhlbi.nih.gov/ga4gh/drs/v1/objects/011c00e5-3449-4cd5-9723-55be4094219b/access/gs', 401, b'<!doctype html>\n\n<html lang=""en"">\n  <head>\n    <meta charset=""utf-8"">\n    <title>401</title>\n    <link href=\'https://fonts.googleapis.com/css?family=Source+Sans+Pro\' rel=\'stylesheet\' type=\'text/css\'>\n  </head>\n\n  <body>\n    <main>\n      <img class=""error-page__logo"" src=""/user/static/img/logo.svg"" />\n      <div class=""error-page"">\n        <div class=""error-page__banner"">\n          <h1>Gen3 Data Commons</h1>\n        </div>\n        <div class=""error-page__status-code"">\n          <div class=""error-page__status-code-text"">\n            <h1>401</h1>\n            <h2>Unauthorized</h2>\n          </div>\n          <img class=""error-page__picture"" src=""/user/static/img/404.svg"" />\n        </div>\n        <div class=""error-page__information"">\n          <p class=""body"">Error Identifier:</p>\n          <p class=""introduction"">36eed37c-1834-4c46-9b0f-e0af9d394e95</p>\n          \n          <p class=""body"">Error Message:</p>\n          <p class=""introduction"">request to arborist failed: error decoding token: illegal base64 data at input byte 0</p>\n          \n          <div>\n            \n            <p class=""body"">Please try again!</p>\n            \n          </div>\n        </div>\n      </div>\n    </main>\n  </body>\n  <style type=""text/css"">\n\n  .body {\n    font-size: 16px;\n    color: #9b9b9b;\n    margin: 0;\n  }\n\n  .introduction {\n    font-size: 18px;\n    font-weight: 500;\n    line-height: 18px;\n    letter-spacing: .03rem;\n    color: #000000;\n    display: inline-block;\n    margin: 13px 0 13px 0;\n  }\n\n  .auth-data-logo {\n    width: 314px;\n    height: 116px;\n    display: block;\n    margin: auto;\n    margin-bottom: 15px;\n  }\n\n  body {\n    margin: 30px 30px;\n    background-color: #F5F5F5;\n    font-family: \'Source Sans Pro\', sans-serif;\n  }\n\n  .error-page {\n    width: 100%;\n    display: flex;\n    flex-flow: column;\n    justify-content: center;\n    background-color: white;\n    border: 1px solid #9b9b9b;\n    border-radius: 8px;\n  }\n\n  .error-page__logo {\n    width: 100%;\n    display: flex;\n    margin-bottom: 15px;\n    max-height: 116px;\n    justify-content: center;\n  }\n\n  .error-page__banner {\n    min-height: 70px;\n    background-color: #3283C8;\n    text-align: center;\n    color: white;\n    border-radius: 8px 8px 0 0;\n    font-weight: lighter;\n  }\n\n  .error-page__status-code {\n    display: flex;\n    justify-content: center;\n    align-items: center;\n    margin: 60px 0 25px 0;\n    text-align: center;\n    background-color: white;\n  }\n\n  .error-page__status-code-text {\n    display: flex;\n    flex-flow: column;\n    margin-right: 20px;\n  }\n\n  .error-page__status-code-text h1 {\n    font-size: 155px;\n    font-weight: lighter;\n    margin: 0;\n    line-height: 1em;\n  }\n\n  .error-page__status-code-text h2 {\n    margin: 0;\n  }\n\n  .error-page__picture {\n    height: 186px;\n    margin-left: 20px;\n  }\n\n  .error-page__information {\n    margin: 0 30px 60px 30px;\n    border: 1px solid #9b9b9b;\n    border-radius: 4px;\n    background-color: #f5f5f5;\n    padding: 20px 20px;\n  }\n\n  .error-page__information a {\n    color: #3283C8;\n    text-decoration: none;\n  }\n\n  </style>\n</html>')
```

**CloudWatch Logs Insights**    
region: us-east-1    
log-group-names: /aws/lambda/azul-service-dev    
data-sources:     
facets:     
start-time: -3600s    
end-time: 0s    
query-string:
  ```
  fields @timestamp, @message
| filter @message like /3471cc03/ # Add request IDs from "service_5xx (api gateway)" query
# | filter @message like /3471cc03|88d486d5|3e22fcf6|a0ec9fff|24808707|92a94b87/ # Add request IDs from "service_5xx (api gateway)" query
| filter @message like /Received|Got 401|azul-service-dev|azul.terra/
| sort @timestamp asc
| limit 1000

  ```
---
| @timestamp | @message |
| --- | --- |
| 2025-12-19 18:38:46.564 | [INFO] 2025-12-19T18:38:46.564Z 3471cc03-5124-4299-b7d5-c51eec4dbc32 azul.chalice Received GET request for &#x27;&#x2f;fetch&#x2f;repository&#x2f;files&#x2f;e3120445-aa13-421d-b4e9-728e5bc8ed1c&#x27;, with {&quot;query&quot;: {&quot;catalog&quot;: &quot;lm2-it&quot;, &quot;version&quot;: &quot;2025-01-07T21:14:38.183927Z&quot;}, &quot;headers&quot;: {&quot;accept-encoding&quot;: &quot;identity&quot;, &quot;cloudfront-forwarded-proto&quot;: &quot;https&quot;, &quot;cloudfront-is-desktop-viewer&quot;: &quot;true&quot;, &quot;cloudfront-is-mobile-viewer&quot;: &quot;false&quot;, &quot;cloudfront-is-smarttv-viewer&quot;: &quot;false&quot;, &quot;cloudfront-is-tablet-viewer&quot;: &quot;false&quot;, &quot;cloudfront-viewer-asn&quot;: &quot;14618&quot;, &quot;cloudfront-viewer-country&quot;: &quot;US&quot;, &quot;host&quot;: &quot;service.dev.singlecell.gi.ucsc.edu&quot;, &quot;user-agent&quot;: &quot;python-urllib3&#x2f;1.26.20&quot;, &quot;via&quot;: &quot;1.1 dfc1931cc62ecd4133c2b9bdae1bb476.cloudfront.net (CloudFront)&quot;, &quot;x-amz-cf-id&quot;: &quot;2rCzhhdNl14Y2Aew8xapa7ABcsRiZvyhtEFPxhWjWJrG4rYRnNPcQA==&quot;, &quot;x-amzn-trace-id&quot;: &quot;Root=1-69459bb6-1837b47508dff47a416ecffd&quot;, &quot;x-forwarded-for&quot;: &quot;52.20.215.168, 15.158.50.182&quot;, &quot;x-forwarded-port&quot;: &quot;443&quot;, &quot;x-forwarded-proto&quot;: &quot;https&quot;}}. |
| 2025-12-19 18:38:46.854 | [INFO] 2025-12-19T18:38:46.854Z 3471cc03-5124-4299-b7d5-c51eec4dbc32 azul.terra Making GET request to &#x27;https:&#x2f;&#x2f;gen3.biodatacatalyst.nhlbi.nih.gov&#x2f;ga4gh&#x2f;drs&#x2f;v1&#x2f;objects&#x2f;011c00e5-3449-4cd5-9723-55be4094219b&#x27; |
| 2025-12-19 18:38:46.854 | [INFO] 2025-12-19T18:38:46.854Z 3471cc03-5124-4299-b7d5-c51eec4dbc32 azul.terra … without a request body |
| 2025-12-19 18:38:46.854 | [INFO] 2025-12-19T18:38:46.854Z 3471cc03-5124-4299-b7d5-c51eec4dbc32 azul.terra … with request headers [(&#x27;x-goog-api-client&#x27;, &#x27;cred-type&#x2f;sa&#x27;), (&#x27;authorization&#x27;, &#x27;REDACTED&#x27;)] |
| 2025-12-19 18:38:46.953 | [INFO] 2025-12-19T18:38:46.953Z 3471cc03-5124-4299-b7d5-c51eec4dbc32 azul.terra Got 200 response after 0.099s from GET to https:&#x2f;&#x2f;gen3.biodatacatalyst.nhlbi.nih.gov&#x2f;ga4gh&#x2f;drs&#x2f;v1&#x2f;objects&#x2f;011c00e5-3449-4cd5-9723-55be4094219b |
| 2025-12-19 18:38:46.953 | [INFO] 2025-12-19T18:38:46.953Z 3471cc03-5124-4299-b7d5-c51eec4dbc32 azul.terra … with response headers HTTPHeaderDict({&#x27;Date&#x27;: &#x27;Fri, 19 Dec 2025 18:38:46 GMT&#x27;, &#x27;Content-Type&#x27;: &#x27;application&#x2f;json&#x27;, &#x27;Content-Length&#x27;: &#x27;847&#x27;, &#x27;Connection&#x27;: &#x27;keep-alive&#x27;, &#x27;Server&#x27;: &#x27;nginx&#x27;, &#x27;X-Frame-Options&#x27;: &#x27;SAMEORIGIN&#x27;, &#x27;X-Content-Type-Options&#x27;: &#x27;nosniff&#x27;, &#x27;X-Xss-Protection&#x27;: &#x27;1; mode=block&#x27;, &#x27;Strict-Transport-Security&#x27;: &#x27;max-age=63072000; includeSubdomains;&#x27;, &#x27;Access-Control-Allow-Origin&#x27;: &#x27;*&#x27;, &#x27;Access-Control-Allow-Methods&#x27;: &#x27;GET, POST, OPTIONS, DELETE, PUT&#x27;, &#x27;Access-Control-Allow-Headers&#x27;: &#x27;DNT,User-Agent,X-Requested-With,If-Modified-Since,Cache-Control,Content-Type,Range,Authorization,Cookie,X-CSRF-Token&#x27;, &#x27;Access-Control-Expose-Headers&#x27;: &#x27;Content-Length,Content-Range&#x27;, &#x27;Set-Cookie&#x27;: &#x27;csrftoken=56fdeeaacc1d6c58a6220271994e2c2014010.0002025-12-19T18:38:46+00:00;Path=&#x2f;;Secure;SameSite=Lax, session=56fdeeaacc1d6c58a6220271994e2c20;Path=&#x2f;;Max-Age=1200;HttpOnly;Secure;SameSite=Lax, visitor=56fdeeaacc1d6c58a6220271994e2c20;Path=&#x2f;;Max-Age=36000000;HttpOnly;Secure;SameSite=Lax&#x27;}) |
| 2025-12-19 18:38:46.953 | [INFO] 2025-12-19T18:38:46.953Z 3471cc03-5124-4299-b7d5-c51eec4dbc32 azul.terra … with a response body of length 847 being b&#x27;{&quot;access_methods&quot;:[{&quot;access_id&quot;:&quot;gs&quot;,&quot;access_url&quot;:{&quot;url&quot;:&quot;gs:&#x2f;&#x2f;nih-nhlbi-bdc-phs001961-v3-c1&#x2f;sequencefiles&#x2f;fdadee7e-2097-45d5-bf81-cc280bd8348e&#x2f;ACD3_snRNA_S1_L001_I1_001.fastq.gz&quot;},&quot;region&quot;:&quot;&quot;,&quot;type&quot;:&quot;gs&quot;},{&quot;access_id&quot;:&quot;s3&quot;,&quot;access_url&quot;:{&quot;url&quot;:&quot;s3:&#x2f;&#x2f;nih-nhlbi-bdc-phs001961-v3-c1&#x2f;sequencefiles&#x2f;fdadee7e-2097-45d5-bf81-cc280bd8348e&#x2f;ACD3_snRNA_S1_L001_I1_001.fastq.gz&quot;},&quot;region&quot;:&quot;&quot;,&quot;type&quot;:&quot;s3&quot;}],&quot;aliases&quot;:[],&quot;checksums&quot;:[{&quot;checksum&quot;:&quot;37b72280438f62206327056a63b15672&quot;,&quot;type&quot;:&quot;md5&quot;}],&quot;created_time&quot;:null,&quot;description&quot;:null,&quot;form&quot;:&quot;object&quot;,&quot;id&quot;:&quot;dg.4503&#x2f;011c00e5-3449-4cd5-9723-55be4094219b&quot;,&quot;index_created_time&quot;:&quot;2025-01-16T22:40:44.385552&quot;,&quot;index_updated_time&quot;:&quot;2025-01-16T22:40:44.385559&quot;,&quot;mime_type&quot;:&quot;application&#x2f;json&quot;,&quot;name&quot;:&quot;&quot;,&quot;self_uri&quot;:&quot;drs:&#x2f;&#x2f;dg.4503:011c00e5-3449-4cd5-9723-55be4094219b&quot;,&quot;size&quot;:209115034,&quot;updated_time&quot;:null,&quot;version&quot;:null}\n&#x27; |
| 2025-12-19 18:38:46.954 | [INFO] 2025-12-19T18:38:46.954Z 3471cc03-5124-4299-b7d5-c51eec4dbc32 azul.terra Making GET request to &#x27;https:&#x2f;&#x2f;gen3.biodatacatalyst.nhlbi.nih.gov&#x2f;ga4gh&#x2f;drs&#x2f;v1&#x2f;objects&#x2f;011c00e5-3449-4cd5-9723-55be4094219b&#x2f;access&#x2f;gs&#x27; |
| 2025-12-19 18:38:46.955 | [INFO] 2025-12-19T18:38:46.955Z 3471cc03-5124-4299-b7d5-c51eec4dbc32 azul.terra … without a request body |
| 2025-12-19 18:38:46.955 | [INFO] 2025-12-19T18:38:46.955Z 3471cc03-5124-4299-b7d5-c51eec4dbc32 azul.terra … with request headers [(&#x27;x-goog-api-client&#x27;, &#x27;cred-type&#x2f;sa&#x27;), (&#x27;authorization&#x27;, &#x27;REDACTED&#x27;)] |
| 2025-12-19 18:38:47.136 | [INFO] 2025-12-19T18:38:47.136Z 3471cc03-5124-4299-b7d5-c51eec4dbc32 azul.terra Got 401 response after 0.181s from GET to https:&#x2f;&#x2f;gen3.biodatacatalyst.nhlbi.nih.gov&#x2f;ga4gh&#x2f;drs&#x2f;v1&#x2f;objects&#x2f;011c00e5-3449-4cd5-9723-55be4094219b&#x2f;access&#x2f;gs |
| 2025-12-19 18:38:47.136 | [INFO] 2025-12-19T18:38:47.136Z 3471cc03-5124-4299-b7d5-c51eec4dbc32 azul.terra … with response headers HTTPHeaderDict({&#x27;Date&#x27;: &#x27;Fri, 19 Dec 2025 18:38:47 GMT&#x27;, &#x27;Content-Type&#x27;: &#x27;text&#x2f;html; charset=utf-8&#x27;, &#x27;Content-Length&#x27;: &#x27;3080&#x27;, &#x27;Connection&#x27;: &#x27;keep-alive&#x27;, &#x27;Server&#x27;: &#x27;nginx&#x27;, &#x27;Access-Control-Expose-Headers&#x27;: &#x27;*, Content-Length,Content-Range&#x27;, &#x27;Set-Cookie&#x27;: &#x27;fence=; Expires=Thu, 01 Jan 1970 00:00:00 GMT; Secure; HttpOnly; Path=&#x2f;, access_token=; Expires=Thu, 01 Jan 1970 00:00:00 GMT; Secure; HttpOnly; Path=&#x2f;; SameSite=Lax&#x27;, &#x27;X-Frame-Options&#x27;: &#x27;SAMEORIGIN&#x27;, &#x27;X-Content-Type-Options&#x27;: &#x27;nosniff&#x27;, &#x27;X-Xss-Protection&#x27;: &#x27;1; mode=block&#x27;, &#x27;Strict-Transport-Security&#x27;: &#x27;max-age=63072000; includeSubdomains;&#x27;, &#x27;Access-Control-Allow-Origin&#x27;: &#x27;*&#x27;, &#x27;Access-Control-Allow-Methods&#x27;: &#x27;GET, POST, OPTIONS, DELETE, PUT&#x27;, &#x27;Access-Control-Allow-Headers&#x27;: &#x27;DNT,User-Agent,X-Requested-With,If-Modified-Since,Cache-Control,Content-Type,Range,Authorization,Cookie,X-CSRF-Token&#x27;}) |
| 2025-12-19 18:38:47.136 | [INFO] 2025-12-19T18:38:47.136Z 3471cc03-5124-4299-b7d5-c51eec4dbc32 azul.terra … with a response body of length 3080 starting in b&#x27;&lt;!doctype html&gt;\n\n&lt;html lang=&quot;en&quot;&gt;\n  &lt;head&gt;\n    &lt;meta charset=&quot;utf-8&quot;&gt;\n    &lt;title&gt;401&lt;&#x2f;title&gt;\n    &lt;link href=\&#x27;https:&#x2f;&#x2f;fonts.googleapis.com&#x2f;css?family=Source+Sans+Pro\&#x27; rel=\&#x27;stylesheet\&#x27; type=\&#x27;text&#x2f;css\&#x27;&gt;\n  &lt;&#x2f;head&gt;\n\n  &lt;body&gt;\n    &lt;main&gt;\n      &lt;img class=&quot;error-page__logo&quot; src=&quot;&#x2f;user&#x2f;static&#x2f;img&#x2f;logo.svg&quot; &#x2f;&gt;\n      &lt;div class=&quot;error-page&quot;&gt;\n        &lt;div class=&quot;error-page__banner&quot;&gt;\n          &lt;h1&gt;Gen3 Data Commons&lt;&#x2f;h1&gt;\n        &lt;&#x2f;div&gt;\n        &lt;div class=&quot;error-page__status-code&quot;&gt;\n          &lt;div class=&quot;error-page__status-code-text&quot;&gt;\n            &lt;h1&gt;401&lt;&#x2f;h1&gt;\n            &lt;h2&gt;Unauthorized&lt;&#x2f;h2&gt;\n          &lt;&#x2f;div&gt;\n          &lt;img class=&quot;error-page__picture&quot; src=&quot;&#x2f;user&#x2f;static&#x2f;img&#x2f;404.svg&quot; &#x2f;&gt;\n        &lt;&#x2f;div&gt;\n        &lt;div class=&quot;error-page__information&quot;&gt;\n          &lt;p class=&quot;body&quot;&gt;Error Identifier:&lt;&#x2f;p&gt;\n          &lt;p class=&quot;introduction&quot;&gt;36eed37c-1834-4c46-9b0f-e0af9d394e95&lt;&#x2f;p&gt;\n          \n          &lt;p class=&quot;body&quot;&gt;Error Message:&lt;&#x2f;p&gt;\n          &lt;p class=&quot;introduction&quot;&gt;request to arborist failed: error decoding token: illegal base64 data...&#x27; |
| 2025-12-19 18:38:47.139 | [ERROR] 2025-12-19T18:38:47.136Z 3471cc03-5124-4299-b7d5-c51eec4dbc32 azul-service-dev Caught exception for path &#x2f;fetch&#x2f;repository&#x2f;files&#x2f;{file_uuid} Traceback (most recent call last):   File &quot;&#x2f;var&#x2f;task&#x2f;chalice&#x2f;app.py&quot;, line 1922, in _get_view_function_response     response = view_function(**function_args)   File &quot;&#x2f;var&#x2f;task&#x2f;app.py&quot;, line 1534, in fetch_repository_files     body = _repository_files(file_uuid, fetch=True)   File &quot;&#x2f;var&#x2f;task&#x2f;app.py&quot;, line 1548, in _repository_files     return app.repository_controller.download_file(catalog=catalog,            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^                                                    fetch=fetch,                                                    ^^^^^^^^^^^^     ...&lt;2 lines&gt;...                                                    headers=headers,                                                    ^^^^^^^^^^^^^^^^                                                    authentication=request.authentication)                                                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;service&#x2f;repository_controller.py&quot;, line 288, in download_file     download.update(plugin, authentication)     ~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;plugins&#x2f;repository&#x2f;tdr.py&quot;, line 274, in update     access = drs_client.get_object(self.file.drs_uri,                                    access_method=AccessMethod.gs)   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;drs.py&quot;, line 362, in get_object     return self._get_object(drs_uri, access_method)            ~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;drs.py&quot;, line 384, in _get_object     return self._get_object_access(drs_uri, access_id, AccessMethod.https)            ~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;drs.py&quot;, line 427, in _get_object_access     raise DRSStatusException(url, response) azul.drs.DRSStatusException: (&#x27;Unexpected response from https:&#x2f;&#x2f;gen3.biodatacatalyst.nhlbi.nih.gov&#x2f;ga4gh&#x2f;drs&#x2f;v1&#x2f;objects&#x2f;011c00e5-3449-4cd5-9723-55be4094219b&#x2f;access&#x2f;gs&#x27;, 401, b&#x27;&lt;!doctype html&gt;\n\n&lt;html lang=&quot;en&quot;&gt;\n  &lt;head&gt;\n    &lt;meta charset=&quot;utf-8&quot;&gt;\n    &lt;title&gt;401&lt;&#x2f;title&gt;\n    &lt;link href=\&#x27;https:&#x2f;&#x2f;fonts.googleapis.com&#x2f;css?family=Source+Sans+Pro\&#x27; rel=\&#x27;stylesheet\&#x27; type=\&#x27;text&#x2f;css\&#x27;&gt;\n  &lt;&#x2f;head&gt;\n\n  &lt;body&gt;\n    &lt;main&gt;\n      &lt;img class=&quot;error-page__logo&quot; src=&quot;&#x2f;user&#x2f;static&#x2f;img&#x2f;logo.svg&quot; &#x2f;&gt;\n      &lt;div class=&quot;error-page&quot;&gt;\n        &lt;div class=&quot;error-page__banner&quot;&gt;\n          &lt;h1&gt;Gen3 Data Commons&lt;&#x2f;h1&gt;\n        &lt;&#x2f;div&gt;\n        &lt;div class=&quot;error-page__status-code&quot;&gt;\n          &lt;div class=&quot;error-page__status-code-text&quot;&gt;\n            &lt;h1&gt;401&lt;&#x2f;h1&gt;\n            &lt;h2&gt;Unauthorized&lt;&#x2f;h2&gt;\n          &lt;&#x2f;div&gt;\n          &lt;img class=&quot;error-page__picture&quot; src=&quot;&#x2f;user&#x2f;static&#x2f;img&#x2f;404.svg&quot; &#x2f;&gt;\n        &lt;&#x2f;div&gt;\n        &lt;div class=&quot;error-page__information&quot;&gt;\n          &lt;p class=&quot;body&quot;&gt;Error Identifier:&lt;&#x2f;p&gt;\n          &lt;p class=&quot;introduction&quot;&gt;36eed37c-1834-4c46-9b0f-e0af9d394e95&lt;&#x2f;p&gt;\n          \n          &lt;p class=&quot;body&quot;&gt;Error Message:&lt;&#x2f;p&gt;\n          &lt;p class=&quot;introduction&quot;&gt;request to arborist failed: error decoding token: illegal base64 data at input byte 0&lt;&#x2f;p&gt;\n          \n          &lt;div&gt;\n            \n            &lt;p class=&quot;body&quot;&gt;Please try again!&lt;&#x2f;p&gt;\n            \n          &lt;&#x2f;div&gt;\n        &lt;&#x2f;div&gt;\n      &lt;&#x2f;div&gt;\n    &lt;&#x2f;main&gt;\n  &lt;&#x2f;body&gt;\n  &lt;style type=&quot;text&#x2f;css&quot;&gt;\n\n  .body {\n    font-size: 16px;\n    color: #9b9b9b;\n    margin: 0;\n  }\n\n  .introduction {\n    font-size: 18px;\n    font-weight: 500;\n    line-height: 18px;\n    letter-spacing: .03rem;\n    color: #000000;\n    display: inline-block;\n    margin: 13px 0 13px 0;\n  }\n\n  .auth-data-logo {\n    width: 314px;\n    height: 116px;\n    display: block;\n    margin: auto;\n    margin-bottom: 15px;\n  }\n\n  body {\n    margin: 30px 30px;\n    background-color: #F5F5F5;\n    font-family: \&#x27;Source Sans Pro\&#x27;, sans-serif;\n  }\n\n  .error-page {\n    width: 100%;\n    display: flex;\n    flex-flow: column;\n    justify-content: center;\n    background-color: white;\n    border: 1px solid #9b9b9b;\n    border-radius: 8px;\n  }\n\n  .error-page__logo {\n    width: 100%;\n    display: flex;\n    margin-bottom: 15px;\n    max-height: 116px;\n    justify-content: center;\n  }\n\n  .error-page__banner {\n    min-height: 70px;\n    background-color: #3283C8;\n    text-align: center;\n    color: white;\n    border-radius: 8px 8px 0 0;\n    font-weight: lighter;\n  }\n\n  .error-page__status-code {\n    display: flex;\n    justify-content: center;\n    align-items: center;\n    margin: 60px 0 25px 0;\n    text-align: center;\n    background-color: white;\n  }\n\n  .error-page__status-code-text {\n    display: flex;\n    flex-flow: column;\n    margin-right: 20px;\n  }\n\n  .error-page__status-code-text h1 {\n    font-size: 155px;\n    font-weight: lighter;\n    margin: 0;\n    line-height: 1em;\n  }\n\n  .error-page__status-code-text h2 {\n    margin: 0;\n  }\n\n  .error-page__picture {\n    height: 186px;\n    margin-left: 20px;\n  }\n\n  .error-page__information {\n    margin: 0 30px 60px 30px;\n    border: 1px solid #9b9b9b;\n    border-radius: 4px;\n    background-color: #f5f5f5;\n    padding: 20px 20px;\n  }\n\n  .error-page__information a {\n    color: #3283C8;\n    text-decoration: none;\n  }\n\n  &lt;&#x2f;style&gt;\n&lt;&#x2f;html&gt;&#x27;) |
---

Which also causes IT `_test_repository_files` to  fail, https://gitlab.dev.singlecell.gi.ucsc.edu/ucsc/azul/-/jobs/131114

```
2025-12-19 18:38:46,532    INFO MainThread test.integration_test: Beginning sub-test [repository_files] {'catalog': 'lm2-it'}
2025-12-19 18:38:46,534    INFO MainThread test.integration_test: Making GET request to 'https://service.dev.singlecell.gi.ucsc.edu/fetch/repository/files/e3120445-aa13-421d-b4e9-728e5bc8ed1c?catalog=lm2-it&version=2025-01-07T21%3A14%3A38.183927Z'
2025-12-19 18:38:46,534    INFO MainThread test.integration_test: … without a request body
2025-12-19 18:38:46,535    INFO MainThread test.integration_test: … without request headers
2025-12-19 18:38:47,152    INFO MainThread test.integration_test: Got 500 response after 0.618s from GET to https://service.dev.singlecell.gi.ucsc.edu/fetch/repository/files/e3120445-aa13-421d-b4e9-728e5bc8ed1c?catalog=lm2-it&version=2025-01-07T21%3A14%3A38.183927Z
2025-12-19 18:38:47,152    INFO MainThread test.integration_test: … with response headers HTTPHeaderDict({'Content-Type': 'application/json', 'Content-Length': '128', 'Connection': 'keep-alive', 'Date': 'Fri, 19 Dec 2025 18:38:47 GMT', 'X-Amzn-Trace-Id': 'Root=1-69459bb6-1837b47508dff47a416ecffd;Parent=5ca1589d23c012e9;Sampled=0;Lineage=1:b3b00478:0', 'x-amzn-RequestId': '3a69349d-3ee3-4a0e-aad8-fd96435b7d9a', 'Referrer-Policy': 'strict-origin-when-cross-origin', 'X-XSS-Protection': '1; mode=block', 'Access-Control-Allow-Origin': '*', 'Strict-Transport-Security': 'max-age=63072000; includeSubDomains; preload', 'Access-Control-Allow-Headers': 'Authorization,Content-Type,X-Amz-Date,X-Amz-Security-Token,X-Api-Key', 'X-Frame-Options': 'DENY', 'Content-Security-Policy': "default-src 'self';img-src 'self' data:;script-src 'self';style-src 'self';frame-ancestors 'none';form-action 'self'", 'x-amz-apigw-id': 'V2VElGdLoAMEQpw=', 'Cache-Control': 'no-store', 'X-Content-Type-Options': 'nosniff', 'X-Cache': 'Error from cloudfront', 'Via': '1.1 dfc1931cc62ecd4133c2b9bdae1bb476.cloudfront.net (CloudFront)', 'X-Amz-Cf-Pop': 'IAD12-P3', 'X-Amz-Cf-Id': '2rCzhhdNl14Y2Aew8xapa7ABcsRiZvyhtEFPxhWjWJrG4rYRnNPcQA=='})
2025-12-19 18:38:47,153    INFO MainThread test.integration_test: … with a response body of length 128 being b'{"Code":"InternalServerError","Message":"An internal server error occurred.","RequestId":"3471cc03-5124-4299-b7d5-c51eec4dbc32"}'
2025-12-19 18:38:47,153    INFO MainThread test.integration_test: Making GET request to 'https://service.dev.singlecell.gi.ucsc.edu/fetch/repository/files/e3120445-aa13-421d-b4e9-728e5bc8ed1c?catalog=lm2-it&version=2025-01-07T21%3A14%3A38.183927Z'
2025-12-19 18:38:47,153    INFO MainThread test.integration_test: … without a request body
2025-12-19 18:38:47,153    INFO MainThread test.integration_test: … without request headers
2025-12-19 18:38:47,413    INFO MainThread test.integration_test: Got 500 response after 0.260s from GET to https://service.dev.singlecell.gi.ucsc.edu/fetch/repository/files/e3120445-aa13-421d-b4e9-728e5bc8ed1c?catalog=lm2-it&version=2025-01-07T21%3A14%3A38.183927Z
2025-12-19 18:38:47,414    INFO MainThread test.integration_test: … with response headers HTTPHeaderDict({'Content-Type': 'application/json', 'Content-Length': '128', 'Connection': 'keep-alive', 'Date': 'Fri, 19 Dec 2025 18:38:47 GMT', 'X-Amzn-Trace-Id': 'Root=1-69459bb7-012590f405ba7087283b205a;Parent=3d1748f81596a6e2;Sampled=0;Lineage=1:b3b00478:0', 'x-amzn-RequestId': 'a4a6be33-7137-4e6d-8e1e-625bd3b0bb4c', 'Referrer-Policy': 'strict-origin-when-cross-origin', 'X-XSS-Protection': '1; mode=block', 'Access-Control-Allow-Origin': '*', 'Strict-Transport-Security': 'max-age=63072000; includeSubDomains; preload', 'Access-Control-Allow-Headers': 'Authorization,Content-Type,X-Amz-Date,X-Amz-Security-Token,X-Api-Key', 'X-Frame-Options': 'DENY', 'Content-Security-Policy': "default-src 'self';img-src 'self' data:;script-src 'self';style-src 'self';frame-ancestors 'none';form-action 'self'", 'x-amz-apigw-id': 'V2VErGxNIAMEfsg=', 'Cache-Control': 'no-store', 'X-Content-Type-Options': 'nosniff', 'X-Cache': 'Error from cloudfront', 'Via': '1.1 dfc1931cc62ecd4133c2b9bdae1bb476.cloudfront.net (CloudFront)', 'X-Amz-Cf-Pop': 'IAD12-P3', 'X-Amz-Cf-Id': '3EVPQN-cAfMjs6i59ABNNmSxn9oLnqjdEFrq96yvWx_0V8W9UOER-w=='})
2025-12-19 18:38:47,414    INFO MainThread test.integration_test: … with a response body of length 128 being b'{"Code":"InternalServerError","Message":"An internal server error occurred.","RequestId":"88d486d5-0004-47f7-906d-107880f55d88"}'
2025-12-19 18:38:47,414    INFO MainThread test.integration_test: Making GET request to 'https://service.dev.singlecell.gi.ucsc.edu/fetch/repository/files/e3120445-aa13-421d-b4e9-728e5bc8ed1c?catalog=lm2-it&version=2025-01-07T21%3A14%3A38.183927Z'
2025-12-19 18:38:47,414    INFO MainThread test.integration_test: … without a request body
2025-12-19 18:38:47,415    INFO MainThread test.integration_test: … without request headers
2025-12-19 18:38:47,631    INFO MainThread test.integration_test: Got 500 response after 0.217s from GET to https://service.dev.singlecell.gi.ucsc.edu/fetch/repository/files/e3120445-aa13-421d-b4e9-728e5bc8ed1c?catalog=lm2-it&version=2025-01-07T21%3A14%3A38.183927Z
2025-12-19 18:38:47,631    INFO MainThread test.integration_test: … with response headers HTTPHeaderDict({'Content-Type': 'application/json', 'Content-Length': '128', 'Connection': 'keep-alive', 'Date': 'Fri, 19 Dec 2025 18:38:47 GMT', 'X-Amzn-Trace-Id': 'Root=1-69459bb7-54e6c94d64765c01518c9a63;Parent=6cf52557e75c0799;Sampled=0;Lineage=1:b3b00478:0', 'x-amzn-RequestId': '23880ade-a91d-478e-ae98-5db5831a8590', 'Referrer-Policy': 'strict-origin-when-cross-origin', 'X-XSS-Protection': '1; mode=block', 'Access-Control-Allow-Origin': '*', 'Strict-Transport-Security': 'max-age=63072000; includeSubDomains; preload', 'Access-Control-Allow-Headers': 'Authorization,Content-Type,X-Amz-Date,X-Amz-Security-Token,X-Api-Key', 'X-Frame-Options': 'DENY', 'Content-Security-Policy': "default-src 'self';img-src 'self' data:;script-src 'self';style-src 'self';frame-ancestors 'none';form-action 'self'", 'x-amz-apigw-id': 'V2VEuHEgIAMEI0Q=', 'Cache-Control': 'no-store', 'X-Content-Type-Options': 'nosniff', 'X-Cache': 'Error from cloudfront', 'Via': '1.1 dfc1931cc62ecd4133c2b9bdae1bb476.cloudfront.net (CloudFront)', 'X-Amz-Cf-Pop': 'IAD12-P3', 'X-Amz-Cf-Id': 'SeO4svCidlCWWXtq3nmIe4hOUpCNQ0ISpfA8FuZ5HVk9kc3t9aaAdg=='})
2025-12-19 18:38:47,631    INFO MainThread test.integration_test: … with a response body of length 128 being b'{"Code":"InternalServerError","Message":"An internal server error occurred.","RequestId":"3e22fcf6-5025-4e7f-a56b-d947c76e0c1e"}'
2025-12-19 18:38:47,632    INFO MainThread test.integration_test: Making GET request to 'https://service.dev.singlecell.gi.ucsc.edu/fetch/repository/files/e3120445-aa13-421d-b4e9-728e5bc8ed1c?catalog=lm2-it&version=2025-01-07T21%3A14%3A38.183927Z'
2025-12-19 18:38:47,632    INFO MainThread test.integration_test: … without a request body
2025-12-19 18:38:47,632    INFO MainThread test.integration_test: … without request headers
2025-12-19 18:38:47,926    INFO MainThread test.integration_test: Got 500 response after 0.294s from GET to https://service.dev.singlecell.gi.ucsc.edu/fetch/repository/files/e3120445-aa13-421d-b4e9-728e5bc8ed1c?catalog=lm2-it&version=2025-01-07T21%3A14%3A38.183927Z
2025-12-19 18:38:47,926    INFO MainThread test.integration_test: … with response headers HTTPHeaderDict({'Content-Type': 'application/json', 'Content-Length': '128', 'Connection': 'keep-alive', 'Date': 'Fri, 19 Dec 2025 18:38:47 GMT', 'X-Amzn-Trace-Id': 'Root=1-69459bb7-595cd38c7af9d6087eb2383d;Parent=5734a4dab4f7a2ab;Sampled=0;Lineage=1:b3b00478:0', 'x-amzn-RequestId': '7d63b5f3-ffa0-4c13-91ea-c2112f20e1a7', 'Referrer-Policy': 'strict-origin-when-cross-origin', 'X-XSS-Protection': '1; mode=block', 'Access-Control-Allow-Origin': '*', 'Strict-Transport-Security': 'max-age=63072000; includeSubDomains; preload', 'Access-Control-Allow-Headers': 'Authorization,Content-Type,X-Amz-Date,X-Amz-Security-Token,X-Api-Key', 'X-Frame-Options': 'DENY', 'Content-Security-Policy': "default-src 'self';img-src 'self' data:;script-src 'self';style-src 'self';frame-ancestors 'none';form-action 'self'", 'x-amz-apigw-id': 'V2VEwFN4oAMEMrw=', 'Cache-Control': 'no-store', 'X-Content-Type-Options': 'nosniff', 'X-Cache': 'Error from cloudfront', 'Via': '1.1 dfc1931cc62ecd4133c2b9bdae1bb476.cloudfront.net (CloudFront)', 'X-Amz-Cf-Pop': 'IAD12-P3', 'X-Amz-Cf-Id': 'iGQho-ZDB7CT_5DJ6dAg5HWbPrBVXbk-Z6qNuolrKaZSSW7s92jEyA=='})
2025-12-19 18:38:47,926    INFO MainThread test.integration_test: … with a response body of length 128 being b'{"Code":"InternalServerError","Message":"An internal server error occurred.","RequestId":"a0ec9fff-6f99-4822-b4f7-542e6ee65202"}'
2025-12-19 18:38:47,927    INFO MainThread test.integration_test: Making GET request to 'https://service.dev.singlecell.gi.ucsc.edu/fetch/repository/files/e3120445-aa13-421d-b4e9-728e5bc8ed1c?catalog=lm2-it&version=2025-01-07T21%3A14%3A38.183927Z'
2025-12-19 18:38:47,927    INFO MainThread test.integration_test: … without a request body
2025-12-19 18:38:47,927    INFO MainThread test.integration_test: … without request headers
2025-12-19 18:38:48,103    INFO MainThread test.integration_test: Got 500 response after 0.176s from GET to https://service.dev.singlecell.gi.ucsc.edu/fetch/repository/files/e3120445-aa13-421d-b4e9-728e5bc8ed1c?catalog=lm2-it&version=2025-01-07T21%3A14%3A38.183927Z
2025-12-19 18:38:48,103    INFO MainThread test.integration_test: … with response headers HTTPHeaderDict({'Content-Type': 'application/json', 'Content-Length': '128', 'Connection': 'keep-alive', 'Date': 'Fri, 19 Dec 2025 18:38:48 GMT', 'X-Amzn-Trace-Id': 'Root=1-69459bb7-09399d8d668dc538516051ce;Parent=526dc65656ec1ac6;Sampled=0;Lineage=1:b3b00478:0', 'x-amzn-RequestId': 'ef57a346-2acc-4842-9789-6de47a7e2050', 'Referrer-Policy': 'strict-origin-when-cross-origin', 'X-XSS-Protection': '1; mode=block', 'Access-Control-Allow-Origin': '*', 'Strict-Transport-Security': 'max-age=63072000; includeSubDomains; preload', 'Access-Control-Allow-Headers': 'Authorization,Content-Type,X-Amz-Date,X-Amz-Security-Token,X-Api-Key', 'X-Frame-Options': 'DENY', 'Content-Security-Policy': "default-src 'self';img-src 'self' data:;script-src 'self';style-src 'self';frame-ancestors 'none';form-action 'self'", 'x-amz-apigw-id': 'V2VEzFI2IAMEH6w=', 'Cache-Control': 'no-store', 'X-Content-Type-Options': 'nosniff', 'X-Cache': 'Error from cloudfront', 'Via': '1.1 dfc1931cc62ecd4133c2b9bdae1bb476.cloudfront.net (CloudFront)', 'X-Amz-Cf-Pop': 'IAD12-P3', 'X-Amz-Cf-Id': 'TRMNoZyn7e7FOAoOSEqwpmt5uT35R54wihpDUiJz_4zBcz_ec-ftQA=='})
2025-12-19 18:38:48,103    INFO MainThread test.integration_test: … with a response body of length 128 being b'{"Code":"InternalServerError","Message":"An internal server error occurred.","RequestId":"24808707-d94e-44d9-be4c-f14764f735d1"}'
2025-12-19 18:38:48,103    INFO MainThread test.integration_test: Making GET request to 'https://service.dev.singlecell.gi.ucsc.edu/fetch/repository/files/e3120445-aa13-421d-b4e9-728e5bc8ed1c?catalog=lm2-it&version=2025-01-07T21%3A14%3A38.183927Z'
2025-12-19 18:38:48,103    INFO MainThread test.integration_test: … without a request body
2025-12-19 18:38:48,104    INFO MainThread test.integration_test: … without request headers
2025-12-19 18:38:48,313    INFO MainThread test.integration_test: Got 500 response after 0.210s from GET to https://service.dev.singlecell.gi.ucsc.edu/fetch/repository/files/e3120445-aa13-421d-b4e9-728e5bc8ed1c?catalog=lm2-it&version=2025-01-07T21%3A14%3A38.183927Z
2025-12-19 18:38:48,313    INFO MainThread test.integration_test: … with response headers HTTPHeaderDict({'Content-Type': 'application/json', 'Content-Length': '128', 'Connection': 'keep-alive', 'Date': 'Fri, 19 Dec 2025 18:38:48 GMT', 'X-Amzn-Trace-Id': 'Root=1-69459bb8-66116d2a118e413637ae91e9;Parent=398b74472f517120;Sampled=0;Lineage=1:b3b00478:0', 'x-amzn-RequestId': '76dfe27c-c76f-4183-b2c2-50a2b6d06914', 'Referrer-Policy': 'strict-origin-when-cross-origin', 'X-XSS-Protection': '1; mode=block', 'Access-Control-Allow-Origin': '*', 'Strict-Transport-Security': 'max-age=63072000; includeSubDomains; preload', 'Access-Control-Allow-Headers': 'Authorization,Content-Type,X-Amz-Date,X-Amz-Security-Token,X-Api-Key', 'X-Frame-Options': 'DENY', 'Content-Security-Policy': "default-src 'self';img-src 'self' data:;script-src 'self';style-src 'self';frame-ancestors 'none';form-action 'self'", 'x-amz-apigw-id': 'V2VE1ECUoAMEptw=', 'Cache-Control': 'no-store', 'X-Content-Type-Options': 'nosniff', 'X-Cache': 'Error from cloudfront', 'Via': '1.1 dfc1931cc62ecd4133c2b9bdae1bb476.cloudfront.net (CloudFront)', 'X-Amz-Cf-Pop': 'IAD12-P3', 'X-Amz-Cf-Id': 'JClcTG199pjXnHEUQX6mbnha-oHONVgblz7RVBIbxCn1s318uJZSlg=='})
2025-12-19 18:38:48,314    INFO MainThread test.integration_test: … with a response body of length 128 being b'{"Code":"InternalServerError","Message":"An internal server error occurred.","RequestId":"92a94b87-d96a-4e60-9827-0925ce67caa0"}'
2025-12-19 18:38:48,314    INFO MainThread test.integration_test: Failed sub-test [repository_files] {'catalog': 'lm2-it'}

  test_indexing (integration_test.IndexingIntegrationTest.test_indexing) [repository_files] (catalog='lm2-it') ... ERROR

…

======================================================================
ERROR: test_indexing (integration_test.IndexingIntegrationTest.test_indexing) [repository_files] (catalog='lm2-it')
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/builds/ucsc/azul/test/integration_test.py", line 385, in subTest
    yield
  File "/builds/ucsc/azul/test/integration_test.py", line 1142, in _test_repository_files
    self._test_file_download(source, inner_file)
    ~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^
  File "/builds/ucsc/azul/test/integration_test.py", line 1154, in _test_file_download
    response = self._get_url_unchecked(GET, file_url)
  File "/builds/ucsc/azul/test/integration_test.py", line 967, in _get_url_unchecked
    response = http.request(method=method,
                            url=url,
    ...<2 lines>...
                            timeout=float(config.api_gateway_lambda_timeout + 1),
                            preload_content=not stream)
  File "/build/.venv/lib/python3.13/site-packages/urllib3/request.py", line 77, in request
    return self.request_encode_url(
           ~~~~~~~~~~~~~~~~~~~~~~~^
        method, url, fields=fields, headers=headers, **urlopen_kw
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    )
    ^
  File "/build/.venv/lib/python3.13/site-packages/urllib3/request.py", line 99, in request_encode_url
    return self.urlopen(method, url, **extra_kw)
           ~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/builds/ucsc/azul/src/azul/http.py", line 446, in urlopen
    raise urllib3.exceptions.MaxRetryError(pool, url)
urllib3.exceptions.MaxRetryError: test.integration_test.LoggingHTTPSConnectionPool(host='service.dev.singlecell.gi.ucsc.edu', port=443): Max retries exceeded with url: https://service.dev.singlecell.gi.ucsc.edu/fetch/repository/files/e3120445-aa13-421d-b4e9-728e5bc8ed1c?catalog=lm2-it&version=2025-01-07T21%3A14%3A38.183927Z (Caused by None)

----------------------------------------------------------------------
Ran 16 tests in 501.155s

FAILED (errors=1, skipped=1)
make: *** [Makefile:271: integration_test] Error 1
section_end:1766169701:step_script
[0Ksection_start:1766169701:cleanup_file_variables
[0K[0K[36;1mCleaning up project directory and file based variables[0;m[0;m
section_end:1766169701:cleanup_file_variables
[0K[31;1mERROR: Job failed: exit code 1[0;m

```

Recurrence: https://gitlab.dev.singlecell.gi.ucsc.edu/ucsc/azul/-/jobs/131881

```
2026-01-03 07:22:55,652    INFO MainThread test.integration_test: Making GET request to 'https://service.dev.singlecell.gi.ucsc.edu/fetch/repository/files/e3120445-aa13-421d-b4e9-728e5bc8ed1c?catalog=lm2-it&version=2025-01-07T21%3A14%3A38.183927Z'
2026-01-03 07:22:55,652    INFO MainThread test.integration_test: … without a request body
2026-01-03 07:22:55,652    INFO MainThread test.integration_test: … without request headers
2026-01-03 07:22:55,823    INFO MainThread test.integration_test: Got 500 response after 0.171s from GET to https://service.dev.singlecell.gi.ucsc.edu/fetch/repository/files/e3120445-aa13-421d-b4e9-728e5bc8ed1c?catalog=lm2-it&version=2025-01-07T21%3A14%3A38.183927Z
2026-01-03 07:22:55,823    INFO MainThread test.integration_test: … with response headers HTTPHeaderDict({'Content-Type': 'application/json', 'Content-Length': '128', 'Connection': 'keep-alive', 'Date': 'Sat, 03 Jan 2026 07:22:55 GMT', 'X-Amzn-Trace-Id': 'Root=1-6958c3cf-78f1a02974234da36bd54d17;Parent=0596e0ec0818dd81;Sampled=0;Lineage=1:b3b00478:0', 'x-amzn-RequestId': 'dd8c62e5-a02c-41ea-bbc4-da8fa43661d0', 'Referrer-Policy': 'strict-origin-when-cross-origin', 'X-XSS-Protection': '1; mode=block', 'Access-Control-Allow-Origin': '*', 'Strict-Transport-Security': 'max-age=63072000; includeSubDomains; preload', 'Access-Control-Allow-Headers': 'Authorization,Content-Type,X-Amz-Date,X-Amz-Security-Token,X-Api-Key', 'X-Frame-Options': 'DENY', 'Content-Security-Policy': "default-src 'self';img-src 'self' data:;script-src 'self';style-src 'self';frame-ancestors 'none';form-action 'self'", 'x-amz-apigw-id': 'WmOIgGHNIAMEZEQ=', 'Cache-Control': 'no-store', 'X-Content-Type-Options': 'nosniff', 'X-Cache': 'Error from cloudfront', 'Via': '1.1 caafbc8a9aa04b09dd564a3ddef60622.cloudfront.net (CloudFront)', 'X-Amz-Cf-Pop': 'IAD12-P3', 'X-Amz-Cf-Id': 'z_yFRXyh_OJyO-vC_suc0o645jNr5mMqBfSGM2jHlAsgVar5pwLm1w=='})
2026-01-03 07:22:55,823    INFO MainThread test.integration_test: … with a response body of length 128 being b'{"Code":"InternalServerError","Message":"An internal server error occurred.","RequestId":"9f9a3c89-83a5-4ea6-95d4-353ad17c3424"}'
2026-01-03 07:22:55,824    INFO MainThread test.integration_test: Failed sub-test [repository_files] {'catalog': 'lm2-it'}

  test_indexing (integration_test.IndexingIntegrationTest.test_indexing) [repository_files] (catalog='lm2-it') ... ERROR
```

Application logs

```
[ERROR] 2026-01-03T07:22:55.815Z  9f9a3c89-83a5-4ea6-95d4-353ad17c3424  azul-service-dev  Caught exception for path /fetch/repository/files/{file_uuid}
Traceback (most recent call last):
  File "/var/task/chalice/app.py", line 1922, in _get_view_function_response
    response = view_function(**function_args)
  File "/var/task/app.py", line 1534, in fetch_repository_files
    body = _repository_files(file_uuid, fetch=True)
  File "/var/task/app.py", line 1548, in _repository_files
    return app.repository_controller.download_file(catalog=catalog,
           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^
                                                   fetch=fetch,
                                                   ^^^^^^^^^^^^
    ...<2 lines>...
                                                   headers=headers,
                                                   ^^^^^^^^^^^^^^^^
                                                   authentication=request.authentication)
                                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/var/task/azul/service/repository_controller.py", line 288, in download_file
    download.update(plugin, authentication)
    ~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^
  File "/var/task/azul/plugins/repository/tdr.py", line 274, in update
    access = drs_client.get_object(self.file.drs_uri,
                                   access_method=AccessMethod.gs)
  File "/var/task/azul/drs.py", line 362, in get_object
    return self._get_object(drs_uri, access_method)
           ~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^
  File "/var/task/azul/drs.py", line 384, in _get_object
    return self._get_object_access(drs_uri, access_id, AccessMethod.https)
           ~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/var/task/azul/drs.py", line 427, in _get_object_access
    raise DRSStatusException(url, response)
azul.drs.DRSStatusException: ('Unexpected response from https://gen3.biodatacatalyst.nhlbi.nih.gov/ga4gh/drs/v1/objects/011c00e5-3449-4cd5-9723-55be4094219b/access/gs', 401, b'<!doctype html>\n\n<html lang="en">…</html>')"
"[INFO] 2026-01-03T07:22:55.817Z  9f9a3c89-83a5-4ea6-95d4-353ad17c3424  azul.chalice  Returning 500 response with headers {"headers": {"Access-Control-Allow-Origin": "*", "Access-Control-Allow-Headers": "Authorization,Content-Type,X-Amz-Date,X-Amz-Security-Token,X-Api-Key", "Content-Security-Policy": "default-src 'self';img-src 'self' data:;script-src 'self';style-src 'self';frame-ancestors 'none';form-action 'self'", "Referrer-Policy": "strict-origin-when-cross-origin", "Strict-Transport-Security": "max-age=63072000; includeSubDomains; preload", "X-Content-Type-Options": "nosniff", "X-Frame-Options": "DENY", "X-XSS-Protection": "1; mode=block", "Cache-Control": "no-store"}}."
"[INFO] 2026-01-03T07:22:55.817Z  9f9a3c89-83a5-4ea6-95d4-353ad17c3424  azul.chalice  … with a response body of length 133 being {"Code": "InternalServerError", "Message": "An internal server error occurred.", "RequestId": "9f9a3c89-83a5-4ea6-95d4-353ad17c3424"}
```

The HTML from the body of the DRS server response in rendered form:

<img width="1702" height="1506" alt="Image" src="https://github.com/user-attachments/assets/2b28ea7c-8591-4466-812a-8266dcc62578" />

Recurrence: https://gitlab.dev.singlecell.gi.ucsc.edu/ucsc/azul/-/jobs/132240

```
2026-01-08 18:46:47,147    INFO MainThread test.integration_test: Making GET request to 'https://service.dev.singlecell.gi.ucsc.edu/fetch/repository/files/e3120445-aa13-421d-b4e9-728e5bc8ed1c?catalog=lm2-it&version=2025-01-07T21%3A14%3A38.183927Z'
2026-01-08 18:46:47,147    INFO MainThread test.integration_test: … without a request body
2026-01-08 18:46:47,148    INFO MainThread test.integration_test: … without request headers
2026-01-08 18:46:47,315    INFO MainThread test.integration_test: Got 500 response after 0.168s from GET to https://service.dev.singlecell.gi.ucsc.edu/fetch/repository/files/e3120445-aa13-421d-b4e9-728e5bc8ed1c?catalog=lm2-it&version=2025-01-07T21%3A14%3A38.183927Z
2026-01-08 18:46:47,315    INFO MainThread test.integration_test: … with response headers HTTPHeaderDict({'Content-Type': 'application/json', 'Content-Length': '128', 'Connection': 'keep-alive', 'Date': 'Thu, 08 Jan 2026 18:46:47 GMT', 'X-Amzn-Trace-Id': 'Root=1-695ffb97-12f1b18944f230e06cb47017;Parent=3eee3cb4c128b57d;Sampled=0;Lineage=1:b3b00478:0', 'x-amzn-RequestId': '69e4f94e-b2ef-493e-bc9a-861242dd0f16', 'Referrer-Policy': 'strict-origin-when-cross-origin', 'X-XSS-Protection': '1; mode=block', 'Access-Control-Allow-Origin': '*', 'Strict-Transport-Security': 'max-age=63072000; includeSubDomains; preload', 'Access-Control-Allow-Headers': 'Authorization,Content-Type,X-Amz-Date,X-Amz-Security-Token,X-Api-Key', 'X-Frame-Options': 'DENY', 'Content-Security-Policy': "default-src 'self';img-src 'self' data:;script-src 'self';style-src 'self';frame-ancestors 'none';form-action 'self'", 'x-amz-apigw-id': 'W4Q_rF_8oAMELNw=', 'Cache-Control': 'no-store', 'X-Content-Type-Options': 'nosniff', 'X-Cache': 'Error from cloudfront', 'Via': '1.1 c5f8f8068a88ebb73e505f5e51b5262e.cloudfront.net (CloudFront)', 'X-Amz-Cf-Pop': 'IAD12-P3', 'X-Amz-Cf-Id': 'ZQJouZX8buIBmv8iGhJQpvNJA6stcUsueyNdb_mZae9oSL7qB2k3Gw=='})
2026-01-08 18:46:47,315    INFO MainThread test.integration_test: … with a response body of length 128 being b'{"Code":"InternalServerError","Message":"An internal server error occurred.","RequestId":"b58bfded-dfbe-4ab2-b72b-b7599d95f213"}'
2026-01-08 18:46:47,316    INFO MainThread test.integration_test: Failed sub-test [repository_files] {'catalog': 'lm2-it'}
  test_indexing (integration_test.IndexingIntegrationTest.test_indexing) [repository_files] (catalog='lm2-it') ... ERROR
```

Application logs

**CloudWatch Logs Insights**    
region: us-east-1    
log-group-names: /aws/lambda/azul-service-dev    
data-sources:     
facets:     
start-time: -3600s    
end-time: 0s    
query-string:
  ```
  fields @timestamp, @message
| filter @message like /42581b07|01e052e3|132afda4|e2ff86f6|334dfbfe|b58bfde/
| filter @message like /Returning|Received|Got 401|azul-service-dev/
| sort @timestamp asc
| limit 4
  ```
---
| @timestamp | @message |
| --- | --- |
| 2026-01-08 18:46:45.616 | [INFO] 2026-01-08T18:46:45.616Z 42581b07-1953-477d-bea6-23e06c535b13 azul.chalice Received GET request for &#x27;&#x2f;fetch&#x2f;repository&#x2f;files&#x2f;e3120445-aa13-421d-b4e9-728e5bc8ed1c&#x27;, with {&quot;query&quot;: {&quot;catalog&quot;: &quot;lm2-it&quot;, &quot;version&quot;: &quot;2025-01-07T21:14:38.183927Z&quot;}, &quot;headers&quot;: {&quot;accept-encoding&quot;: &quot;identity&quot;, &quot;cloudfront-forwarded-proto&quot;: &quot;https&quot;, &quot;cloudfront-is-desktop-viewer&quot;: &quot;true&quot;, &quot;cloudfront-is-mobile-viewer&quot;: &quot;false&quot;, &quot;cloudfront-is-smarttv-viewer&quot;: &quot;false&quot;, &quot;cloudfront-is-tablet-viewer&quot;: &quot;false&quot;, &quot;cloudfront-viewer-asn&quot;: &quot;14618&quot;, &quot;cloudfront-viewer-country&quot;: &quot;US&quot;, &quot;host&quot;: &quot;service.dev.singlecell.gi.ucsc.edu&quot;, &quot;user-agent&quot;: &quot;python-urllib3&#x2f;1.26.20&quot;, &quot;via&quot;: &quot;1.1 c5f8f8068a88ebb73e505f5e51b5262e.cloudfront.net (CloudFront)&quot;, &quot;x-amz-cf-id&quot;: &quot;4BZGh2BN2jz94bDJ4fzXSB7lLojd-UAiDRwJ0oRKoaFrJvDnBAvG2A==&quot;, &quot;x-amzn-trace-id&quot;: &quot;Root=1-695ffb95-7890ad7e10ef07f06509ecb0&quot;, &quot;x-forwarded-for&quot;: &quot;52.20.215.168, 15.158.50.183&quot;, &quot;x-forwarded-port&quot;: &quot;443&quot;, &quot;x-forwarded-proto&quot;: &quot;https&quot;}}. |
| 2026-01-08 18:46:46.274 | [INFO] 2026-01-08T18:46:46.274Z 42581b07-1953-477d-bea6-23e06c535b13 azul.terra Got 401 response after 0.122s from GET to https:&#x2f;&#x2f;gen3.biodatacatalyst.nhlbi.nih.gov&#x2f;ga4gh&#x2f;drs&#x2f;v1&#x2f;objects&#x2f;011c00e5-3449-4cd5-9723-55be4094219b&#x2f;access&#x2f;gs |
| 2026-01-08 18:46:46.277 | [ERROR] 2026-01-08T18:46:46.274Z 42581b07-1953-477d-bea6-23e06c535b13 azul-service-dev Caught exception for path &#x2f;fetch&#x2f;repository&#x2f;files&#x2f;{file_uuid} Traceback (most recent call last):   File &quot;&#x2f;var&#x2f;task&#x2f;chalice&#x2f;app.py&quot;, line 1922, in _get_view_function_response     response = view_function(**function_args)   File &quot;&#x2f;var&#x2f;task&#x2f;app.py&quot;, line 1534, in fetch_repository_files     body = _repository_files(file_uuid, fetch=True)   File &quot;&#x2f;var&#x2f;task&#x2f;app.py&quot;, line 1548, in _repository_files     return app.repository_controller.download_file(catalog=catalog,            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^                                                    fetch=fetch,                                                    ^^^^^^^^^^^^     ...&lt;2 lines&gt;...                                                    headers=headers,                                                    ^^^^^^^^^^^^^^^^                                                    authentication=request.authentication)                                                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;service&#x2f;repository_controller.py&quot;, line 288, in download_file     download.update(plugin, authentication)     ~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;plugins&#x2f;repository&#x2f;tdr.py&quot;, line 274, in update     access = drs_client.get_object(self.file.drs_uri,                                    access_method=AccessMethod.gs)   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;drs.py&quot;, line 362, in get_object     return self._get_object(drs_uri, access_method)            ~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;drs.py&quot;, line 384, in _get_object     return self._get_object_access(drs_uri, access_id, AccessMethod.https)            ~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;drs.py&quot;, line 427, in _get_object_access     raise DRSStatusException(url, response) azul.drs.DRSStatusException: (&#x27;Unexpected response from https:&#x2f;&#x2f;gen3.biodatacatalyst.nhlbi.nih.gov&#x2f;ga4gh&#x2f;drs&#x2f;v1&#x2f;objects&#x2f;011c00e5-3449-4cd5-9723-55be4094219b&#x2f;access&#x2f;gs&#x27;, 401, b&#x27;&lt;!doctype html&gt;\n\n&lt;html lang=&quot;en&quot;&gt;\n  &lt;head&gt;\n    &lt;meta charset=&quot;utf-8&quot;&gt;\n    &lt;title&gt;401&lt;&#x2f;title&gt;\n    &lt;link href=\&#x27;https:&#x2f;&#x2f;fonts.googleapis.com&#x2f;css?family=Source+Sans+Pro\&#x27; rel=\&#x27;stylesheet\&#x27; type=\&#x27;text&#x2f;css\&#x27;&gt;\n  &lt;&#x2f;head&gt;\n\n  &lt;body&gt;\n    &lt;main&gt;\n      &lt;img class=&quot;error-page__logo&quot; src=&quot;&#x2f;user&#x2f;static&#x2f;img&#x2f;logo.svg&quot; &#x2f;&gt;\n      &lt;div class=&quot;error-page&quot;&gt;\n        &lt;div class=&quot;error-page__banner&quot;&gt;\n          &lt;h1&gt;Gen3 Data Commons&lt;&#x2f;h1&gt;\n        &lt;&#x2f;div&gt;\n        &lt;div class=&quot;error-page__status-code&quot;&gt;\n          &lt;div class=&quot;error-page__status-code-text&quot;&gt;\n            &lt;h1&gt;401&lt;&#x2f;h1&gt;\n            &lt;h2&gt;Unauthorized&lt;&#x2f;h2&gt;\n          &lt;&#x2f;div&gt;\n          &lt;img class=&quot;error-page__picture&quot; src=&quot;&#x2f;user&#x2f;static&#x2f;img&#x2f;404.svg&quot; &#x2f;&gt;\n        &lt;&#x2f;div&gt;\n        &lt;div class=&quot;error-page__information&quot;&gt;\n          &lt;p class=&quot;body&quot;&gt;Error Identifier:&lt;&#x2f;p&gt;\n          &lt;p class=&quot;introduction&quot;&gt;26bd52a8-06ae-45b4-86db-c16f1a4cd74f&lt;&#x2f;p&gt;\n          \n          &lt;p class=&quot;body&quot;&gt;Error Message:&lt;&#x2f;p&gt;\n          &lt;p class=&quot;introduction&quot;&gt;request to arborist failed: error decoding token: illegal base64 data at input byte 0&lt;&#x2f;p&gt;\n          \n          &lt;div&gt;\n            \n            &lt;p class=&quot;body&quot;&gt;Please try again!&lt;&#x2f;p&gt;\n            \n          &lt;&#x2f;div&gt;\n        &lt;&#x2f;div&gt;\n      &lt;&#x2f;div&gt;\n    &lt;&#x2f;main&gt;\n  &lt;&#x2f;body&gt;\n  &lt;style type=&quot;text&#x2f;css&quot;&gt;\n\n  .body {\n    font-size: 16px;\n    color: #9b9b9b;\n    margin: 0;\n  }\n\n  .introduction {\n    font-size: 18px;\n    font-weight: 500;\n    line-height: 18px;\n    letter-spacing: .03rem;\n    color: #000000;\n    display: inline-block;\n    margin: 13px 0 13px 0;\n  }\n\n  .auth-data-logo {\n    width: 314px;\n    height: 116px;\n    display: block;\n    margin: auto;\n    margin-bottom: 15px;\n  }\n\n  body {\n    margin: 30px 30px;\n    background-color: #F5F5F5;\n    font-family: \&#x27;Source Sans Pro\&#x27;, sans-serif;\n  }\n\n  .error-page {\n    width: 100%;\n    display: flex;\n    flex-flow: column;\n    justify-content: center;\n    background-color: white;\n    border: 1px solid #9b9b9b;\n    border-radius: 8px;\n  }\n\n  .error-page__logo {\n    width: 100%;\n    display: flex;\n    margin-bottom: 15px;\n    max-height: 116px;\n    justify-content: center;\n  }\n\n  .error-page__banner {\n    min-height: 70px;\n    background-color: #3283C8;\n    text-align: center;\n    color: white;\n    border-radius: 8px 8px 0 0;\n    font-weight: lighter;\n  }\n\n  .error-page__status-code {\n    display: flex;\n    justify-content: center;\n    align-items: center;\n    margin: 60px 0 25px 0;\n    text-align: center;\n    background-color: white;\n  }\n\n  .error-page__status-code-text {\n    display: flex;\n    flex-flow: column;\n    margin-right: 20px;\n  }\n\n  .error-page__status-code-text h1 {\n    font-size: 155px;\n    font-weight: lighter;\n    margin: 0;\n    line-height: 1em;\n  }\n\n  .error-page__status-code-text h2 {\n    margin: 0;\n  }\n\n  .error-page__picture {\n    height: 186px;\n    margin-left: 20px;\n  }\n\n  .error-page__information {\n    margin: 0 30px 60px 30px;\n    border: 1px solid #9b9b9b;\n    border-radius: 4px;\n    background-color: #f5f5f5;\n    padding: 20px 20px;\n  }\n\n  .error-page__information a {\n    color: #3283C8;\n    text-decoration: none;\n  }\n\n  &lt;&#x2f;style&gt;\n&lt;&#x2f;html&gt;&#x27;) |
| 2026-01-08 18:46:46.278 | [INFO] 2026-01-08T18:46:46.278Z 42581b07-1953-477d-bea6-23e06c535b13 azul.chalice Returning 500 response with headers {&quot;headers&quot;: {&quot;Access-Control-Allow-Origin&quot;: &quot;*&quot;, &quot;Access-Control-Allow-Headers&quot;: &quot;Authorization,Content-Type,X-Amz-Date,X-Amz-Security-Token,X-Api-Key&quot;, &quot;Content-Security-Policy&quot;: &quot;default-src &#x27;self&#x27;;img-src &#x27;self&#x27; data:;script-src &#x27;self&#x27;;style-src &#x27;self&#x27;;frame-ancestors &#x27;none&#x27;;form-action &#x27;self&#x27;&quot;, &quot;Referrer-Policy&quot;: &quot;strict-origin-when-cross-origin&quot;, &quot;Strict-Transport-Security&quot;: &quot;max-age=63072000; includeSubDomains; preload&quot;, &quot;X-Content-Type-Options&quot;: &quot;nosniff&quot;, &quot;X-Frame-Options&quot;: &quot;DENY&quot;, &quot;X-XSS-Protection&quot;: &quot;1; mode=block&quot;, &quot;Cache-Control&quot;: &quot;no-store&quot;}}. |
---

--- a/src/azul/drs.py
+++ b/src/azul/drs.py
@@ -34,7 +34,6 @@ import urllib3.request
 from azul import (
     R,
     cache,
-    cached_property,
     mutable_furl,
 )
 from azul.http import (
@@ -199,14 +198,6 @@ class DRSURI(metaclass=ABCMeta):
             else:
                 raise
 
-    @abstractmethod
-    def to_url(self, client: 'DRSClient', access_id: str | None = None) -> furl:
-        """
-        Translate this DRS URI into a URL of the DRS REST API endpoint at which
-        the file identified by this DRS URI can be accessed.
-        """
-        raise NotImplementedError
-
 
 @attr.s(auto_attribs=True, kw_only=True, frozen=True, slots=True)
 class HostBasedDRSURI(DRSURI):
@@ -228,8 +219,8 @@ class HostBasedDRSURI(DRSURI):
         assert len(path) == 1, R('Invalid path', drs_uri)
         return cls(server=parsed_uri.netloc, object_id=path[0])
 
-    def to_url(self, client: 'DRSClient', access_id: str | None = None) -> furl:
-        path = drs_object_url_path(object_id=self.object_id, access_id=access_id)
+    def to_url(self) -> furl:
+        path = drs_object_url_path(object_id=self.object_id)
         return furl(scheme='https', netloc=self.server, path=path)
 
 
@@ -281,23 +272,23 @@ class CompactDRSURI(DRSURI):
     def _decode(cls, s: str) -> str:
         return urllib.parse.unquote(s, errors='strict')
 
-    def to_url(self, client: 'DRSClient', access_id: str | None = None) -> furl:
+    def to_url(self, id_client: 'IdentifiersDotOrgClient') -> furl:
         if self.provider_code is not None:
             raise NotImplementedError(
                 'Resolving compact identifier-based DRS URIs with '
                 'provider codes is currently not supported', self
             )
-        url = client.id_client.resolve(self.namespace, self.accession)
+        url = id_client.resolve(self.namespace, self.accession)
         # The URL pattern registered at identifiers.org ought to replicate the
-        # DRS spec, but we have to re-create the path using the spec because the
-        # registered pattern does not support embedding the access ID.
+        # DRS spec. If the response to a request to the returned URL includes an
+        # access ID, another request must be made to the returned URL followed
+        # by the string `/access/` and the ID.
         assert str(url.path) == drs_object_url_path(object_id=self.accession), R(
             'Format of resolved URL is incompatible with the DRS specification', url)
-        url.set(path=drs_object_url_path(object_id=self.accession, access_id=access_id))
         return url
 
 
-class IdentifiersDotOrgClient(HasCachedHttpClient):
+class _BaseClient(HasCachedHttpClient):
 
     def _create_http_client(self) -> urllib3.request.RequestMethods:
         return Propagate429HttpClient(
@@ -305,6 +296,26 @@ class IdentifiersDotOrgClient(HasCachedHttpClient):
                 super()._create_http_client()
             )
         )
+
+
+class DRSClient(metaclass=ABCMeta):
+
+    @abstractmethod
+    def drs_object(self, drs_url: furl) -> 'DRSObject':
+        raise NotImplementedError
+
+
+class UnauthenticatedDRSClient(DRSClient, _BaseClient):
+    """
+    A generic DRS client that does not send authentication to the server.
+    """
+
+    def drs_object(self, drs_url: furl) -> 'DRSObject':
+        return DRSObject(url=drs_url,
+                         http_client=self._http_client)
+
+
+class IdentifiersDotOrgClient(_BaseClient):
 
     def resolve(self, prefix: str, accession: str) -> mutable_furl:
         namespace_id = self._prefix_to_namespace(prefix)
@@ -343,26 +354,20 @@ class IdentifiersDotOrgClient(HasCachedHttpClient):
 
 
 @attr.s(auto_attribs=True, kw_only=True, frozen=True)
-class DRSClient:
+class DRSObject:
     _http_client: urllib3.request.RequestMethods
+    _url: furl
 
-    @cached_property
-    def id_client(self) -> IdentifiersDotOrgClient:
-        return IdentifiersDotOrgClient()
-
-    def get_object(self,
-                   drs_uri: str,
-                   access_method: AccessMethod = AccessMethod.https
-                   ) -> Access:
+    def get(self, access_method: AccessMethod = AccessMethod.https) -> Access:
         """
         Returns access to the content of the data object identified by the
         given URI. The scheme of the URL in the returned access object depends
         on the access method specified.
         """
-        return self._get_object(drs_uri, access_method)
+        return self._get(access_method)
 
-    def _get_object(self, drs_uri: str, access_method: AccessMethod) -> Access:
-        url = DRSURI.parse(drs_uri).to_url(self)
+    def _get(self, access_method: AccessMethod) -> Access:
+        url = self._url
         while True:
             response = self._request(url)
             if response.status == 200:
@@ -381,9 +386,9 @@ class DRSClient:
                     # https://github.com/ga4gh/data-repository-service-schemas/issues/361
                     assert access_method is AccessMethod.gs, R(
                         'Unexpected access method', access_method)
-                    return self._get_object_access(drs_uri, access_id, AccessMethod.https)
+                    return self._get_access(access_id, AccessMethod.https)
                 elif access_id is not None:
-                    return self._get_object_access(drs_uri, access_id, access_method)
+                    return self._get_access(access_id, access_method)
                 elif access_url is not None:
                     scheme = furl(access_url['url']).scheme
                     assert scheme == access_method.scheme, R(
@@ -400,12 +405,9 @@ class DRSClient:
             else:
                 raise DRSStatusException(url, response)
 
-    def _get_object_access(self,
-                           drs_uri: str,
-                           access_id: str,
-                           access_method: AccessMethod
-                           ) -> Access:
-        url = DRSURI.parse(drs_uri).to_url(self, access_id)
+    def _get_access(self, access_id: str, access_method: AccessMethod) -> Access:
+        url = self._url.copy()
+        url.path.add(['access', access_id])
         while True:
             response = self._request(url)
             if response.status == 200:

--- a/src/azul/indexer/mirror_service.py
+++ b/src/azul/indexer/mirror_service.py
@@ -722,8 +722,8 @@ class MirrorService(BaseMirrorService, HasCachedHttpClient):
             'Only TDR catalogs are supported', self.catalog)
         assert file.drs_uri is not None, R(
             'File cannot be downloaded', file)
-        drs = self._repository_plugin.drs_client(authentication=None)
-        access = drs.get_object(file.drs_uri, AccessMethod.gs)
+        object = self._repository_plugin.drs_object(file.drs_uri)
+        access = object.get(AccessMethod.gs)
         assert access.method is AccessMethod.https, access
         return furl(access.url)
 

--- a/src/azul/plugins/__init__.py
+++ b/src/azul/plugins/__init__.py
@@ -23,6 +23,9 @@ from typing import (
 )
 
 import attrs
+from furl import (
+    furl,
+)
 from more_itertools import (
     one,
 )
@@ -43,7 +46,12 @@ from azul.digests import (
     Digest,
 )
 from azul.drs import (
-    DRSClient,
+    CompactDRSURI,
+    DRSObject,
+    DRSURI,
+    HostBasedDRSURI,
+    IdentifiersDotOrgClient,
+    UnauthenticatedDRSClient,
 )
 from azul.indexer import (
     Bundle,
@@ -830,16 +838,36 @@ class RepositoryPlugin[BUNDLE: Bundle = Bundle[SourcedBundleFQID],
 
         raise NotImplementedError
 
-    @abstractmethod
-    def drs_client(self,
+    def drs_object(self,
+                   drs_uri: str,
                    authentication: Authentication | None = None
-                   ) -> DRSClient:
+                   ) -> DRSObject:
         """
         Returns a DRS client that uses the given authentication with requests to
         the DRS server. If a concrete subclass doesn't support authentication,
         it should assert that the argument is ``None``.
         """
-        raise NotImplementedError
+        assert authentication is None, type(authentication)
+        drs_url = self._resolve_drs_uri(drs_uri)
+        return self._unauthenticated_drs.drs_object(drs_url)
+
+    def _resolve_drs_uri(self, drs_uri: str) -> furl:
+        drs_uri = DRSURI.parse(drs_uri)
+        if isinstance(drs_uri, CompactDRSURI):
+            drs_url = drs_uri.to_url(self._identifiers_dot_org)
+        elif isinstance(drs_uri, HostBasedDRSURI):
+            drs_url = drs_uri.to_url()
+        else:
+            assert False
+        return drs_url
+
+    @cached_property
+    def _unauthenticated_drs(self) -> UnauthenticatedDRSClient:
+        return UnauthenticatedDRSClient()
+
+    @cached_property
+    def _identifiers_dot_org(self) -> IdentifiersDotOrgClient:
+        return IdentifiersDotOrgClient()
 
     @abstractmethod
     def file_download_class(self) -> type['RepositoryFileDownload']:

--- a/src/azul/plugins/repository/canned/__init__.py
+++ b/src/azul/plugins/repository/canned/__init__.py
@@ -27,9 +27,6 @@ from azul import (
 from azul.auth import (
     Authentication,
 )
-from azul.drs import (
-    DRSClient,
-)
 from azul.http import (
     HasCachedHttpClient,
 )
@@ -275,12 +272,6 @@ class Plugin(RepositoryPlugin[
 
     def file_download_class(self) -> type[RepositoryFileDownload]:
         return CannedFileDownload
-
-    def drs_client(self,
-                   authentication: Authentication | None = None
-                   ) -> DRSClient:
-        assert authentication is None, type(authentication)
-        return DRSClient(http_client=self._http_client)
 
     def validate_version(self, version: str) -> None:
         parse_dcp2_version(version)

--- a/src/azul/plugins/repository/dss/__init__.py
+++ b/src/azul/plugins/repository/dss/__init__.py
@@ -32,9 +32,6 @@ from azul.collections import (
 from azul.deployment import (
     aws,
 )
-from azul.drs import (
-    DRSClient,
-)
 from azul.http import (
     HasCachedHttpClient,
 )
@@ -197,12 +194,6 @@ class Plugin(RepositoryPlugin[
         url.path.add(['files', file_uuid])
         url.query.add(adict(version=file_version, replica=replica, token=token))
         return str(url)
-
-    def drs_client(self,
-                   authentication: Authentication | None = None
-                   ) -> DRSClient:
-        assert authentication is None, type(authentication)
-        return DRSClient(http_client=self._http_client)
 
     def file_download_class(self) -> type[RepositoryFileDownload]:
         return DSSFileDownload

--- a/src/azul/plugins/repository/tdr.py
+++ b/src/azul/plugins/repository/tdr.py
@@ -23,6 +23,7 @@ from furl import (
 
 from azul import (
     cache_per_thread,
+    config,
     require,
 )
 from azul.auth import (
@@ -35,7 +36,7 @@ from azul.bigquery import (
 )
 from azul.drs import (
     AccessMethod,
-    DRSClient,
+    DRSObject,
 )
 from azul.indexer import (
     Bundle,
@@ -177,13 +178,6 @@ class TDRPlugin[TDR_BUNDLE: TDRBundle,
                                   type(authentication))
         return tdr
 
-    @classmethod
-    @cache_per_thread
-    def _drs_client(cls,
-                    authentication: Authentication | None = None
-                    ) -> DRSClient:
-        return cls._user_authenticated_tdr(authentication).drs_client()
-
     def _lookup_source_id(self, spec: TDRSourceSpec) -> str:
         return self.tdr.lookup_source(spec)
 
@@ -209,10 +203,19 @@ class TDRPlugin[TDR_BUNDLE: TDRBundle,
     def _emulate_bundle(self, bundle_fqid: TDRBundleFQID) -> TDR_BUNDLE:
         raise NotImplementedError
 
-    def drs_client(self,
+    def drs_object(self,
+                   drs_uri: str,
                    authentication: Authentication | None = None
-                   ) -> DRSClient:
-        return self._drs_client(authentication)
+                   ) -> DRSObject:
+        drs_url = self._resolve_drs_uri(drs_uri)
+        tdr_url = config.tdr_service_url
+        # Authenticate only if the DRS server is TDR so that we don't leak user
+        # or service account tokens to untrusted servers.
+        if (drs_url.scheme, drs_url.host) == (tdr_url.scheme, tdr_url.host):
+            drs_client = self._user_authenticated_tdr(authentication)
+        else:
+            drs_client = self._unauthenticated_drs
+        return drs_client.drs_object(drs_url)
 
     def file_download_class(self) -> type[RepositoryFileDownload]:
         return TDRFileDownload
@@ -270,9 +273,8 @@ class TDRFileDownload(RepositoryFileDownload):
             assert self.location is None, self
             assert self.retry_after is None, self
         else:
-            drs_client = plugin.drs_client(authentication)
-            access = drs_client.get_object(self.file.drs_uri,
-                                           access_method=AccessMethod.gs)
+            drs_client = plugin.drs_object(self.file.drs_uri, authentication)
+            access = drs_client.get(access_method=AccessMethod.gs)
             require(access.method is AccessMethod.https, access.method)
             require(access.headers is None, access.headers)
             signed_url = access.url

--- a/src/azul/service/download_controller.py
+++ b/src/azul/service/download_controller.py
@@ -16,6 +16,7 @@ from chalice import (
     BadRequestError,
     NotFoundError,
     TooManyRequestsError,
+    UnauthorizedError,
 )
 
 from azul import (
@@ -32,6 +33,9 @@ from azul.chalice import (
 )
 from azul.collections import (
     adict,
+)
+from azul.drs import (
+    DRSStatusException,
 )
 from azul.http import (
     LimitedTimeoutException,
@@ -168,6 +172,12 @@ class DownloadController(ServiceAppController):
             raise ServiceUnavailableError(*e.args)
         except TooManyRequestsException as e:
             raise TooManyRequestsError(*e.args)
+        except DRSStatusException as e:
+            msg, status, data = e.args
+            if status == UnauthorizedError.STATUS_CODE:
+                raise UnauthorizedError(msg)
+            else:
+                raise
         if download.retry_after is not None:
             retry_after = min(download.retry_after, int(1.3 ** request_index))
             if wait is not None:

--- a/src/azul/terra.py
+++ b/src/azul/terra.py
@@ -77,6 +77,7 @@ from azul.deployment import (
 )
 from azul.drs import (
     DRSClient,
+    DRSObject,
 )
 from azul.http import (
     LimitedRetryHttpClient,
@@ -368,7 +369,7 @@ class SAMClient(TerraClient):
         return self.credentials_provider.insufficient_access(resource)
 
 
-class TDRClient(SAMClient):
+class TDRClient(SAMClient, DRSClient):
     """
     A client for the Broad Institute's Terra Data Repository aka "Jade".
     """
@@ -641,8 +642,8 @@ class TDRClient(SAMClient):
         else:
             return self
 
-    def drs_client(self) -> DRSClient:
-        return DRSClient(http_client=self._http_client)
+    def drs_object(self, drs_url: furl) -> DRSObject:
+        return DRSObject(url=drs_url, http_client=self._http_client)
 
     def get_duos(self,
                  source: TDRSourceRef

--- a/test/integration_test.py
+++ b/test/integration_test.py
@@ -1209,13 +1209,13 @@ class IndexingIntegrationTest(IntegrationTestCase):
                   file: JSON
                   ) -> None:
         repository_plugin = self.azul_client.repository_plugin(catalog)
-        drs = repository_plugin.drs_client()
         file_uuid = lookup(file, 'document_id', 'uuid')
+        drs_uri = f'drs://{config.api_lambda_domain("service")}/{file_uuid}'
+        drs_object = repository_plugin.drs_object(drs_uri)
         for access_method in AccessMethod:
             with self.subTest('drs', catalog=catalog, access_method=AccessMethod.https):
                 log.info('Resolving file %r with DRS using %r', file_uuid, access_method)
-                drs_uri = f'drs://{config.api_lambda_domain("service")}/{file_uuid}'
-                access = drs.get_object(drs_uri, access_method=access_method)
+                access = drs_object.get(access_method)
                 self.assertIsNone(access.headers)
                 if access.method is AccessMethod.https:
                     response = self._get_url(GET, furl(access.url), stream=True)

--- a/test/integration_test.py
+++ b/test/integration_test.py
@@ -1145,24 +1145,31 @@ class IndexingIntegrationTest(IntegrationTestCase):
                 self.assertIsNone(file_url, inner_file)
                 self.assertEqual('lungmap', config.catalogs[catalog].atlas, inner_file)
 
-    def _test_file_download(self, source: SourceSpec, file: JSON) -> mutable_furl:
+    def _test_file_download(self, source: SourceSpec, file: JSON) -> mutable_furl | None:
         file_url = furl(file['azul_url'])
         # FIXME: Use _check_endpoint() instead
         #        https://github.com/DataBiosphere/azul/issues/7373
         self.assertEqual(file_url.path.segments[0], 'repository')
         file_url.path.segments.insert(0, 'fetch')
         response = self._get_url_unchecked(GET, file_url)
-        self.assertEqual(200, response.status)
-        response = json.loads(response.data)
-        while response['Status'] != 302:
-            self.assertEqual(301, response['Status'])
+        if response.status == 401:
+            msg = json.loads(response.data)['Message']
+            prefix = 'Unexpected response from '
+            self.assertEqual(prefix, msg[:len(prefix)])
+            self.assertNotIn(str(config.tdr_service_url), msg)
+            return None
+        else:
+            self.assertEqual(200, response.status)
+            response = json.loads(response.data)
+            while response['Status'] != 302:
+                self.assertEqual(301, response['Status'])
+                self.assertNotIn('Retry-After', response)
+                response = self._get_url_json(GET, furl(response['Location']))
             self.assertNotIn('Retry-After', response)
-            response = self._get_url_json(GET, furl(response['Location']))
-        self.assertNotIn('Retry-After', response)
-        final_file_url = furl(response['Location'])
-        response = self._get_url(GET, final_file_url, stream=True)
-        self._validate_file_response(response, source, file)
-        return final_file_url
+            final_file_url = furl(response['Location'])
+            response = self._get_url(GET, final_file_url, stream=True)
+            self._validate_file_response(response, source, file)
+            return final_file_url
 
     def _file_ext(self, file: JSON) -> str:
         # We believe that the file extension is a more reliable indicator than
@@ -1795,6 +1802,7 @@ class IndexingIntegrationTest(IntegrationTestCase):
                             aws.mirror_bucket, '_it', 'file', f'{digest.value}.{digest.type}',
                         ])
                         actual_url = self._test_file_download(source.spec, file_response)
+                        self.assertIsNotNone(actual_url)
                         actual_url.set(args=None)
                         self.assertEqual(expected_url, actual_url)
             _delete()

--- a/test/service/test_repository_files.py
+++ b/test/service/test_repository_files.py
@@ -43,7 +43,7 @@ from azul.deployment import (
 from azul.drs import (
     Access,
     AccessMethod,
-    DRSClient,
+    DRSObject,
 )
 from azul.http import (
     http_client,
@@ -168,10 +168,8 @@ class TestRepositoryFilesWithTDR(DCP2TestCase, RepositoryFilesTestCase):
                                              'X-Goog-SignedHeaders': 'host',
                                              'X-Goog-Signature': 'SOMESIGNATURE',
                                          })
-                    with mock.patch.object(DRSClient,
-                                           'get_object',
-                                           return_value=Access(method=AccessMethod.https,
-                                                               url=str(pre_signed_gs))):
+                    access = Access(method=AccessMethod.https, url=str(pre_signed_gs))
+                    with mock.patch.object(DRSObject, 'get', return_value=access):
                         response = client.request('GET', str(azul_url), redirect=False)
                         self.assertEqual(200 if fetch else 302, response.status)
                         if fetch:

